greymatter-io/operator
# Support HTTP/S and Basic Auth for target GitOps repository

[sc-33225]

I have made the operator able to specify a GitOps target with HTTP/S and Basic Auth, for customers who want to avoid SSH.

## How to test

I have tested this against an instance of GitLab with HTTPS and Basic Auth launched with elest.io. Except for the note below, it sufficed to specify an HTTPS URL as the `-repo` flag to the operator:

```
args: [
  "-repo", "https://gitlab-z7y6q-u687.vm.elestio.app/danielpcox/greymatter-core.git",
  "-branch", "main",
]
```

and set the necessarily environment variables for auth (pulling the password from a Secret):

```
env: [
  {name: "GITOPS_BASICAUTH_USERNAME", value: "danielpcox"},
  {
    name: "GITOPS_BASICAUTH_PASSWORD",
    valueFrom: {
      secretKeyRef: {
        name: "operator-gitops-basicauth-password"
        key: "password"
        optional: false
      }
    }
  }
]
```

This repository in GitLab isn't a straight mirror of the one in GitHub - it is modified to avoid submodules. (See below.)

You should be able to reproduce my successful installation of a mesh by building this branch and using the above configuration to launch the operator. The actual command I used to launch the operator, given these modifications to greymatter-core (and from inside that repository) was

```
kubectl create namespace gm-operator

kubectl create secret generic operator-gitops-basicauth-password \
  --from-literal=password='REDACTED-API-TOKEN-FROM-GITLAB' \
  -n gm-operator

kubectl create secret docker-registry gm-docker-secret \
  --docker-server=greymatter.jfrog.io \
  --docker-username=$JFROG_USERNAME \
  --docker-password=$JFROG_PASSWORD \
  -n gm-operator

# This is unused by necessary if you don't want to go remove the reference to this secret
kubectl create secret generic greymatter-sync-secret \
  --from-file=ssh-private-key=$HOME/.ssh/gat-stage-stage \
  -n gm-operator

# This refers to my test build of this branch
cue eval -c ./k8s/outputs --out text -t operator_image=greymatter.jfrog.io/internal-oci/operator:daniel \
         -e operator_manifests_yaml | kubectl apply -f -
)

```

## Note on submodules:

Don't use them if you can help it. Support for submodules is shoddy in git-go, but as long as our release bundle doesn't use them, that complication doesn't come into play. I had two problems while trying to work with them:

1. An SSH URL in a submodule fails when the GitOps target top-level repository uses HTTPS - the SSH Agent isn't started, so the submodule clone fails (and of course mixing them is dubious to begin with).
2. Even an HTTPS URL in a submodule can fail if the certificate is somehow untrusted. This turns out to be pretty likely - elest.io, where I tested this, created a certificate that didn't match the path required by a file of the submodule, so the clone failed.
